### PR TITLE
BUG: GH #12223, GH #15262. Allow ints for names in MultiIndex

### LINKED ISSUE #12223: multi-index display bug
Thanks for great tool!

When printing a long DataFrame with multi-index which are result of  set_index(..., append=True),
displaying index values look strange.

``` py
import pandas as pd

def func(n):
    x = [
        [['a1-%d' % i, 'a2-%d' % i] for i in range(n)],
        [['b1-%d' % i, 'b2-%d' % i] for i in range(n)],
    ]
    y = pd.concat([pd.DataFrame(a) for a in x], keys=['X', 'Y']).set_index(1, append=True)
    print(y)

func(5) # no problem
func(50) # bug!
```

When DataFrames are short (in this case, 10 rows), there is no problem.
However, when DataFrames is long (in this case, 100 rows) and omitted middle-part on printing, some columns of index is not correct. 

```
             0
    1         
X 0 a2-0  a1-0
  1 a2-1  a1-1
  2 a2-2  a1-2
  3 a2-3  a1-3
  4 a2-4  a1-4
Y 0 b2-0  b1-0
  1 b2-1  b1-1
  2 b2-2  b1-2
  3 b2-3  b1-3
  4 b2-4  b1-4
                   0
        1           
X a2-0  a2-0    a1-0
  a2-1  a2-1    a1-1
  a2-2  a2-2    a1-2
  a2-3  a2-3    a1-3
  a2-4  a2-4    a1-4
  a2-5  a2-5    a1-5
  a2-6  a2-6    a1-6
  a2-7  a2-7    a1-7
  a2-8  a2-8    a1-8
  a2-9  a2-9    a1-9
...              ...
Y b2-40 b2-40  b1-40
  b2-41 b2-41  b1-41
  b2-42 b2-42  b1-42
  b2-43 b2-43  b1-43
  b2-44 b2-44  b1-44
  b2-45 b2-45  b1-45
  b2-46 b2-46  b1-46
  b2-47 b2-47  b1-47
  b2-48 b2-48  b1-48
  b2-49 b2-49  b1-49

[100 rows x 1 columns]
```

result of pd.show_versions() is as follows.

```
INSTALLED VERSIONS
------------------
commit: None
python: 3.5.1.final.0
python-bits: 64
OS: Linux
OS-release: 2.6.32-279.22.1.el6.x86_64
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.17.1
nose: 1.3.7
pip: 7.1.2
setuptools: 18.5
Cython: 0.23.4
numpy: 1.10.1
scipy: 0.16.0
statsmodels: 0.6.1
IPython: 4.0.1
sphinx: 1.3.1
patsy: 0.4.0
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.4
matplotlib: 1.5.0
openpyxl: 2.2.6
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.7.7
lxml: 3.4.4
bs4: 4.4.1
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.9
pymysql: None
psycopg2: None
Jinja2: None
```

### PR DESCRIPTION
**Pull request by Dr-Irv**
- [X] closes #12223 and #15262
 - [X] tests added / passed
  - `pandas/tests/frame/test_combine_concat.py::TestDataFrameConcatCommon::test_concat_numerical_names`
 - [X] passes ``git diff upstream/master | flake8 --diff``
 - [X] whatsnew entry

See discussion in #10461 .  Now if we internally know we need to call `get_level_values` for a specific level number, there is an internal function for doing that, which should allow better support for when people create indices with names that are integers.

### CODE REVIEW

**Comment by jreback** [comment]
did u mean to close?
this looks pretty reasonable

**Comment by Dr-Irv** [comment]
@jreback yes.  The conflicts indicated you had made a change that was incompatible with what I did, so I need to fix that.  Will create a new pull request when ready.

**Comment by jreback** [comment]
@Dr-Irv you don't need to close a PR to fix that though. simply rebase and force push. you *never* need to open a new PR for the same exact issue.

**Comment by Dr-Irv** [comment]
@jreback OK, didn't know that.  Should I reopen this one when I'm ready?

**Comment by jreback** [comment]
sure

**Comment by jreback** [comment]
see more detail here: http://pandas.pydata.org/pandas-docs/stable/contributing.html#contributing-your-changes-to-pandas

**Comment by jorisvandenbossche** [comment]
@Dr-Irv Already reopened it, because if you push again while the PR is closed, you cannot reopen it anymore.

File: doc/source/whatsnew/v0.20.0.txt [hunk 1]
```diff
@@ -185,11 +185,12 @@ Other enhancements
 - ``Series/DataFrame.asfreq()`` have gained a ``fill_value`` parameter, to fill missing values (:issue:`3715`).
 - ``Series/DataFrame.resample.asfreq`` have gained a ``fill_value`` parameter, to fill missing values during resampling (:issue:`3715`).
 - ``pandas.tools.hashing`` has gained a ``hash_tuples`` routine, and ``hash_pandas_object`` has gained the ability to hash a ``MultiIndex`` (:issue:`15224`)
-- ``Series/DataFrame.squeeze()`` have gained the ``axis`` parameter. (:issue:`15339`)
+- ``Series/DataFrame.squeeze()`` have gained the ``axis`` parameter. (:issue:`15339`)<<<<<<< f4edb053e17e51e8c2bed7c16755c4f7f3222117
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: merge residual :>

File: doc/source/whatsnew/v0.20.0.txt [hunk 2]
```diff
@@ -185,11 +185,12 @@ Other enhancements
...
 - ``DataFrame.to_excel()`` has a new ``freeze_panes`` parameter to turn on Freeze Panes when exporting to Excel (:issue:`15160`)
 - HTML table output skips ``colspan`` or ``rowspan`` attribute if equal to 1. (:issue:`15403`)
 - ``pd.TimedeltaIndex`` now has a custom datetick formatter specifically designed for nanosecond level precision (:issue:`8711`)
 - ``pd.types.concat.union_categoricals`` gained the ``ignore_ordered`` argument to allow ignoring the ordered attribute of unioned categoricals (:issue:`13410`). See the :ref:`categorical union docs <categorical.union>` for more information.
+- Using numerical names in ``MultiIndex`` causes less errors. (:issue:`12223`) (:issue:`15262`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: say instead about the bug report about output formatting with a MI under certain conditions.

Comment (jreback) [inline doc/source/whatsnew/v0.20.0.txt]: ``(:issue:`12223`, :issue:`15262`)``

File: pandas/indexes/base.py [hunk 3]
```diff
@@ -2352,6 +2352,11 @@ def get_level_values(self, level):
         self._validate_index_level(level)
         return self
 
+    def _get_level_values(self, num):
+        # Used to mirror implementation for MultiIndex
```
Comment (jreback) [inline pandas/indexes/base.py]: better to add an actual doc-string

File: pandas/indexes/multi.py [hunk 4]
```diff
@@ -856,6 +856,7 @@ def _get_level_values(self, level):
         Parameters
         ----------
         level : int level
+        copy  : bool whether copy of results should be done
```
Comment (jreback) [inline pandas/indexes/multi.py]: why are you adding this? this is a whole different ball game.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/indexes/multi.py (hunk 4 above)
Comment (Dr-Irv) [inline pandas/indexes/multi.py]: @jreback What I needed `_get_level_values()` to do is have the same behavior as the public `get_level_values()`,  with the assumption of the int argument, so the `copy` argument makes that happen by doing the shallow copy there.

When I first looked at this, there was no `_get_level_values()`, but that got introduced within the past 2 weeks, so I then had to make everything compatible.

Comment (jreback) [inline pandas/indexes/multi.py]: still its not clear why you would actually be making this change, it just adds too much complexity.

show an example of why you think you need it (or simply take it out)

Comment (Dr-Irv) [inline pandas/indexes/multi.py]: @jreback I'll try an alternate implementation and let you review that.

**Comment by Dr-Irv** [comment]
@jreback the last commit I did should have addressed the changes you requested, and we're all green, so let me know if there is more to do

**Comment by jreback** [comment]
thanks @Dr-Irv

**Comment by jorisvandenbossche** [comment]
@Dr-Irv Thanks!

**Pull request closed without merging**